### PR TITLE
Path 2D module

### PR DESCRIPTION
**Pull request by utenma**
As result of https://github.com/TheSpyder/rescript-webapi/issues/132

Path2d module added with following js bindings, see on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Path2D#instance_methods)
- addPath
- arc
- arcTo
- bezierCurveTo
- closePath
- ellipse
- lineTo
- moveTo
- quadraticCurveTo
- rect
- roundRect

Syntax
```rescript
open Webapi.Canvas

let path1 = Path2d.make()
let path2 = Path2d.make(~d="M10 10 h 80 v 80 h -80 Z", ())

path1->Path2d.addPath(path2)

path1->Path2d.arc(~x=1.,~y =2.,~r=3.,~a1 = 4., ~a2 =5.,())
```

Changelog

* added Path2d module
* added Canvas2d ellipse roundRect bindings

Breaking
* Updated common Path2d/Canvas2d params names to match mdn docs
```diff
- ctx->quadraticCurveTo(~cp1x=1., ~cp1y=1., ~x=1., ~y=1.)
+ ctx->quadraticCurveTo(~cpx=1., ~cpy=1., ~x=1., ~y=1.)

- ctx->arc(~x=1., ~y=1., ~r=4., ~startAngle=1., ~endAngle=3., ~anticw =true)
+ ctx->arc(~x=1., ~y=1., ~r=4., ~startAngle=1., ~endAngle=3., ~counterClockWise=true, ())
```
* Added Path2d.make, Dropped Canvas2d.strokePath2D
```diff
- let path: path2d = newPath2D("M24.85,10.126c2.018-4.783,6.628-8.125,11.99-8.125c7.223")
+ let path = Path2d.make(~d="M24.85,10.126c2.018-4.783,6.628-8.125,11.99-8.125c7.223")
```
* Canvas2d functions taking Path2d renamed
```diff
- ctx->strokePath2D(path)
- ctx->fillPath2D(path)
+ ctx->strokePath2d(path1)
+ ctx->fillPath2d(path2)
```

TODO
 - [x] add Path2d module
 - [x] export Path2d from `src/Webapi/Webapi__Canvas.res`
 - [x] remove `Canvas2d.newPath2D()` in favor of `Path2d.make(~d=?)`
 - [x] add missing Canvas2d methods
 - [x] Updated canvas2s test
 - [x] extract common methods between Path2d and Context2D (Functor with parametric `t`)
 - [x] list breaking changes

### CODE REVIEW

**Comment by utenma** [comment]
@TheSpyder what do you do with the generated js files from tests? are they used besides compile check?

**Comment by TheSpyder** [comment]
@utenma sorry I haven't looked at this yet. The "tests" are indeed a compile check; it ensures changes made do not impact the compiled JS (unless that's intentional).

I hope to one day have full library coverage, if you could add a few extra test calls to methods you've added that would be great 👍 

Should we delete the old binding methods you're replacing?

File: src/Webapi/Canvas/Webapi__Canvas__Path_Common.res [hunk 1]
```diff
@@ -0,0 +1,81 @@
+module type ModuleWithT = {
```
Comment (TheSpyder) [inline src/Webapi/Canvas/Webapi__Canvas__Path_Common.res:1]: The style used in this project is `Impl` and `T.t`, with the module type and module in one declaration

https://github.com/TheSpyder/rescript-webapi/blob/87270bd837a826c4ba29c7a8fe91ce7ebf04813a/src/Webapi/Dom/Webapi__Dom__Event.res#L1-L5

https://github.com/TheSpyder/rescript-webapi/blob/18d9d553201d575fa8e194d3370298b556b28747/src/Webapi/Dom/Webapi__Dom__PointerEvent.res#L4-L6

File: tests/Webapi/Canvas/Webapi__Canvas__Canvas2d__test.res [hunk 2]
```diff
@@ -112,11 +112,17 @@ ctx->fillRect(~x=1., ~y=0., ~w=10., ~h=10.)
 ctx->strokeRect(~x=1., ~y=0., ~w=10., ~h=10.)
 ctx->clearRect(~x=1., ~y=0., ~w=10., ~h=10.)
 
-let path: path2d = newPath2D("M24.85,10.126c2.018-4.783,6.628-8.125,11.99-8.125c7.223"
+
+let path1 = Path2d.make()
+let path2 = Path2d.make(~d="M24.85,10.126c2.018-4.783,6.628-8.125,11.99-8.125c7.223"
```
Comment (TheSpyder) [inline tests/Webapi/Canvas/Webapi__Canvas__Canvas2d__test.res:117]: Path2d has a new module now, might as well give it a separate test file?

**Review by TheSpyder — CHANGES_REQUESTED** [review]
Getting there! just needs a few tweaks.

**Comment by utenma** [comment]
@TheSpyder updated list of breaking changes

File: tests/Webapi/Canvas/Webapi__Canvas__Canvas2d__test.res [hunk 3]
```diff
@@ -111,12 +123,3 @@ ctx->strokeText("hi", ~x=1., ~y=0., ())
...
 ctx->clearRect(~x=1., ~y=0., ~w=10., ~h=10.)
-
-let path: path2d = newPath2D("M24.85,10.126c2.018-4.783,6.628-8.125,11.99-8.125c7.223"
-  ++ ",0,12.425,6.179,13.079,13.543 c0,0,0.353,1.828-0.424,5.119c-1.058,4.482"
-  ++ "-3.545,8.464-6.898,11.503L24.85,48L7.402,32.165c-3.353-3.038-5.84-7.021"
-  ++ "-6.898-11.503 c-0.777-3.291-0.424-5.119-0.424-5.119C0.734,8.179,5.936,2"
-  ++ ",13.159,2C18.522,2,22.832,5.343,24.85,10.126z")
-ctx->beginPath
-ctx->strokePath2D(path)
-ctx->fillPath2D(path)
```
Comment (TheSpyder) [inline tests/Webapi/Canvas/Webapi__Canvas__Canvas2d__test.res:122]: These don't have an equivalent test in your new code 🤔

**Review by TheSpyder — APPROVED** [review]
I am truly sorry for letting this sit for over 2 months. I had a long work trip and then things just kept coming up that I had to deal with first.

Just one little tweak and then I'll merge+release it.

File: tests/Webapi/Canvas/Webapi__Canvas__Canvas2d__test.res (hunk 3 above)
Comment (TheSpyder) [reply tests/Webapi/Canvas/Webapi__Canvas__Canvas2d__test.res:122]: oh `beginPath` is already there, and the others moved. My bad.

**Comment by TheSpyder** [comment]
Published as 0.10.0

**Pull request merged**